src/Http/Requests/FrontendFormRequest.php

```diff
@@ -0,0 +1,176 @@
+<?php
+
+namespace Statamic\Http\Requests;
+
+use Illuminate\Contracts\Validation\Validator;
+use Illuminate\Foundation\Http\FormRequest;
+use Illuminate\Support\Facades\URL;
+use Illuminate\Support\Traits\Localizable;
+use Illuminate\Validation\ValidationException;
+use Statamic\Facades\Site;
+
+class FrontendFormRequest extends FormRequest
+{
+    use Localizable;
+
+    private $assets = [];
+    private $cachedFields;
+    private $cachedValidator;
+
+    /**
+     * Get any assets in the request
+     */
+    public function assets(): array
+    {
+        return $this->assets;
+    }
+
+    /**
+     * Determine if the user is authorized to make this request.
+     */
+    public function authorize(): bool
+    {
+        return true;
+    }
+
+    /**
+     * Get the validation messages
+     */
+    public function messages(): array
+    {
+        $site = Site::findByUrl(URL::previous()) ?? Site::default();
+
+        $messages = [];
+        $this->withLocale($site->lang(), function () use (&$messages) {
+            $messages = $this->getFormFields()
+                ->validator()
+                ->validator()
+                ->messages()
+                ->getMessages();
+        });
+
+        return collect($messages)->flatten(1)->all();
+    }
+
+    /**
+     * Get the validation attributes that apply to the request.
+     */
+    public function attributes()
+    {
+        $attributes = $this->getFormFields()->preProcessValidatables()->all()->reduce(function ($carry, $field) {
+            return $carry->merge($field->validationAttributes());
+        }, collect())->all();
+
+        return $attributes;
+    }
+
+    /**
+     * Get the validation rules that apply to the request.
+     */
+    public function rules(): array
+    {
+        return $this->getCustomValidator()->rules();
+    }
+
+    /**
+     * Optionally override the redirect url based on the presence of _error_redirect
+     */
+    protected function getRedirectUrl()
+    {
+        $url = $this->redirector->getUrlGenerator();
+
+        if ($redirect = request()->input('_error_redirect')) {
+            return $url->to($redirect);
+        }
+
+        return $url->previous();
+    }
+
+    protected function failedValidation(Validator $validator)
+    {
+        if (request()->ajax()) {
+            $errors = $validator->errors();
+
+            $response = response([
+                'errors' => $errors->all(),
+                'error' => collect($errors->messages())->map(function ($errors, $field) {
+                    return $errors[0];
+                })->all(),
+            ], 400);
+
+            throw (new ValidationException($validator, $response));
+        }
+
+        return parent::failedValidation($validator);
+    }
+
+    protected function getValidatorInstance()
+    {
+        return $this->getCustomValidator()
+            ->withMessages($this->messages())
+            ->validator();
+    }
+
+    public function extraRules($fields)
```

Comment:
I think you can avoid having the extra rules separated from the validator.

In the `validator` method mentioned in the other comment, you can probably just provide the extra rules right in there.